### PR TITLE
Auto ban

### LINKED ISSUE #2979: Implement auto ban if user triggers anti cheat too many times

### PR DESCRIPTION
**Pull request by Miodec**
Auto bans users if they trigger the anti cheat too many times too quickly. I know variable names are not the best so open to suggetions.

Closes #2979

### CODE REVIEW

**Comment by Bruception** [comment]
Need to include unit tests to improve coverage

File: backend/src/dal/user.ts [hunk 1]
```diff
@@ -595,3 +595,36 @@ export async function removeFavoriteQuote(
...
 }
+
+export async function recordAutoBanEvent(
+  uid: string,
+  maxCount: number,
+  maxHours: number
+): Promise<void> {
+  const user = await getUser(uid, "record auto ban event");
+
+  const SECONDS_PER_HOUR = 3600;
```
[the remark this answers is not in the document]
Comment (Miodec) [reply backend/src/dal/user.ts:608]: Can do

File: backend/src/dal/user.ts [hunk 2]
```diff
@@ -595,3 +595,36 @@ export async function removeFavoriteQuote(
...
+  //remove any old events
+  autoBan = autoBan.filter(
+    (timestamp) => timestamp < Date.now() - maxHours * SECONDS_PER_HOUR
+  );
+
+  //push new event
+  autoBan.push(Date.now());
+
+  //update user, ban if needed
+  const updateObj: {
```
Comment (Bruception) [inline backend/src/dal/user.ts:623]: Why not just do:

```ts
const userUpdates: Partial<MonkeyTypes.User> = {
    autoBan,
};
```

File: backend/__tests__/dal/user.spec.ts [hunk 3]
```diff
@@ -186,4 +187,67 @@ describe("UserDal", () => {
       5
     );
   });
+
+  it("autoBan should automatically ban after configured anticheat triggers", async () => {
+    // given
+    const testUser = {
```
Comment (Bruception) [inline backend/__tests__/dal/user.spec.ts:193]: Maybe we should just define this at the top of the file and reuse.

Comment (Miodec) [reply backend/__tests__/dal/user.spec.ts:193]: Unrelated. Can refactor tests in a different PR.

File: backend/__tests__/dal/user.spec.ts [hunk 4]
```diff
@@ -186,4 +187,67 @@ describe("UserDal", () => {
...
+    await recordAutoBanEvent(testUser.uid, 2, 1);
+    await recordAutoBanEvent(testUser.uid, 2, 1);
+
+    Date.now = jest.fn(() => 36000000);
+
+    await recordAutoBanEvent(testUser.uid, 2, 1);
+
+    // then
+    const updatedUser = await getUser(testUser.uid, "test");
+    expect(updatedUser.banned).toBe(undefined);
```
Comment (Bruception) [inline backend/__tests__/dal/user.spec.ts:251]: Maybe we can add another check for these tests to make sure that timestamps are being properly removed.

**Review by Bruception — APPROVED** [review]
LGTM IIW

**Pull request merged**